umbilical/revocation/checker.go

```diff
@@ -0,0 +1,324 @@
+// certificates on demand. It either uses OCSP or CRLs. The latter are
+// cached on disk.
+package revocation
+
+import (
+	"bytes"
+	"crypto/x509"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"math/big"
+	"net/http"
+	"sync"
+	"time"
+
+	"golang.org/x/crypto/ocsp"
+
+	bolt "go.etcd.io/bbolt"
+)
+
+type Config struct {
+	// Path to file to use to cache
+	Cache string
+}
+
+type crlEntry struct {
+	Expires  time.Time
+	BucketID uint64
+}
+
+type Checker struct {
+	cache *bolt.DB
+
+	fetchMux sync.Mutex
+	// URL -> Cond to wait on a CRL that's currently fetching.
+	crlFetching map[string]*sync.Cond
+}
+
+func NewChecker(cfg Config) (*Checker, error) {
+	var (
+		ret Checker
+		err error
+	)
+
+	ret.cache, err = bolt.Open(cfg.Cache, 0600, nil)
+	if err != nil {
+		return nil, fmt.Errorf("bolt.Open(%s): %w", cfg.Cache, err)
+	}
+
+	err = ret.cache.Update(func(tx *bolt.Tx) error {
+		_, err := tx.CreateBucketIfNotExists([]byte("crls"))
+		return err
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to create crls bucket: %w", err)
+	}
+
+	ret.crlFetching = make(map[string]*sync.Cond)
+
+	return &ret, nil
+}
+
+// Checks whether the given certificate is revoked by first trying OCSP,
+// and then checking CRL.
+//
+// Warning: make sure you trust the issuer and checked the chain.
+// Does not check the signature of the issuer.
+func (c *Checker) Revoked(cert, issuer *x509.Certificate) (
+	bool, error) {
+	if len(cert.OCSPServer) != 0 {
+		return c.revokedOCSP(cert, issuer)
+	}
+
+	if len(cert.CRLDistributionPoints) != 0 {
+		return c.revokedCRL(cert, issuer)
+	}
+
+	return true, errors.New("No revocation mechanism available")
+}
+
+func sendOCSP(url string, req []byte, cert, issuer *x509.Certificate) (
+	*ocsp.Response, error) {
+	// TODO Support GET. It might be slightly faster and is easier to cache.
+	resp, err := http.Post(
+		url,
+		"application/ocsp-request",
+		bytes.NewBuffer(req),
+	)
+
+	if err != nil {
+		return nil, fmt.Errorf("POST: %w", err)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != 200 {
+		return nil, fmt.Errorf("HTTP status %d", resp.StatusCode)
+	}
+
+	bs, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("reading response: %w", err)
+	}
+
+	return ocsp.ParseResponseForCert(bs, cert, issuer)
+}
+
+// TODO Cache OCSP. We can use the resp.NextUpdate field.
+func (c *Checker) revokedOCSP(cert, issuer *x509.Certificate) (bool, error) {
+	req, err := ocsp.CreateRequest(cert, issuer, nil)
+	if err != nil {
+		return true, fmt.Errorf("ocsp.CreateRequest: %w", err)
+	}
+
+	ok := false
+	var (
+		resp *ocsp.Response
+		errs []error
+	)
+
+	// Although not specified in any standard as far as I know, it seems
+	// common to try the OCSP servers in the order they are listed.
+	for _, url := range cert.OCSPServer {
+		resp, err = sendOCSP(url, req, cert, issuer)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		ok = true
```

Comment:
Don't we want to `break` here when there's a successful response? Currently, we continue to loop through all available servers, even if the first server responds.

Reply:
Whoops. Fixed.